### DISCUSSION TITLE
How to build on macOS with self-signed cert?

### DISCUSSION THREAD

**Discussion by jamischarles** [neovide/neovide#1370] [Q&A]
Hey folks!

On my work laptop, they've installed their own self signed root cert.

This leads to the `tls connection init failed: invalid peer certificate contents: invalid peer certificate: UnknownIssuer)` error.

In node/npm I can solve this via an env variable that points at the self signed cert:
`NODE_EXTRA_CA_CERTS=path_to_company_cert.pem npm install`

With cargo it seems that you can do the same with:
`CARGO_HTTP_CAINFO=path_to_company_cert.pem`

BUT, my error happens in a build script for `skia-bindings`.

```
 --- stderr
  thread 'main' panicked at 'Failed to download https://codeload.github.com/rust-skia/skia/tar.gz/m100-0.48.7 (https://codeload.github.com/rust-skia/skia/tar.gz/m100-0.48.7: Connection Failed: tls connection init failed: invalid peer certificate contents: invalid peer certificate: UnknownIssuer)', /Users/jacharles/.cargo/registry/src/github.com-1ecc6299db9ec823/skia-bindings-0.49.1/build_support/binary_cache/download.rs:67:35
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
     1: core::panicking::panic_fmt
               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
     2: build_script_build::build_support::binary_cache::download::download_dependencies::{{closure}}
               at ./build_support/binary_cache/download.rs:67:35
     3: core::result::Result<T,E>::unwrap_or_else
               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/result.rs:1464:23
     4: build_script_build::build_support::binary_cache::download::download_dependencies
               at ./build_support/binary_cache/download.rs:66:23
     5: build_script_build::build_support::binary_cache::download::resolve_dependencies
               at ./build_support/binary_cache/download.rs:17:9
     6: build_script_build::build_support::skia::config::build
               at ./build_support/skia/config.rs:348:9
     7: build_script_build::build_from_source
               at ./build.rs:39:5
     8: build_script_build::main
               at ./build.rs:127:13
     9: core::ops::function::FnOnce::call_once
               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Any ideas for how I can tell rust to use the company cert here?

Thanks!

**Comment by MultisampledNight** [comment]
Probably you're not looking to tell Rust directly to use your cert, but rather `skia-bindings`' build script (skia is a dependency of Neovide). [`skia-bindings` uses `ureq` to make that download request](https://github.com/rust-skia/rust-skia/blob/master/skia-bindings/build_support/binary_cache/utils.rs#L6), which uses in turn `rustls`. But there's a dead end, an open issue since October 2021: https://github.com/rustls/rustls-native-certs/issues/22, I'm sorry.

**Comment by znd4** [comment]
I actually managed to work through what seems to be this exact issue and wrote up the workaround in [an issue](https://github.com/rust-skia/rust-skia/issues/685) on rust-skia. Making it more specific for neovide, it'd be something like:


1. Using something that has ssl certs properly set up, download the binary release for your platform from https://github.com/rust-skia/skia-binaries/releases (in my case, [skia-binaries-ab8bc4dfa8fc85713027-aarch64-apple-darwin-gl.tar.gz](https://github.com/rust-skia/skia-binaries/releases/download/0.53.0/skia-binaries-ab8bc4dfa8fc85713027-aarch64-apple-darwin-gl.tar.gz) from 0.53.0)
2. in a separate terminal window/tab, navigate to your downloads folder and `python -m http.server`
    1. I had to say yes to an OS popup here. Thanks for all of the security apple! (*slight sarcasm*)
4. in your clone of `neovide`, point `SKIA_BINARIES_URL` to your http fileserver and run the cargo steps that failed earlier:

```shell
export SKIA_BINARIES_URL=http://localhost:8000/name-of-the-file-you-downloaded.tar.gz
cargo ... (install steps here depend on OS)
```
